releaseNotes/Microsoft.Featuremanagement.md

```diff
@@ -2,6 +2,45 @@
 
 [Source code][source_code] | [Package (NuGet)][package] | [Samples][samples] | [Product documentation][docs]
 
+## 4.0.0 - September 24, 2024
+
+### Enhancements
+
+#### Variant Feature Flags
+
+A variant feature flag is an enhanced feature flag that supports multiple states or variations. While it can still be toggled on or off, it also allows for different configurations, ranging from simple primitives to complex JSON objects. Variant feature flags are particularly useful for feature rollouts, configuration rollouts, and feature experimentation (also known as A/B testing).
+
+The new `IVariantFeatureManager` has been introduced as the successor to the existing `IFeatureManager`. It retains all the functionalities of `IFeatureManager` while adding new `GetVariantAsync` methods and supporting `CancellationToken` for all methods.
+
+``` C#
+IVariantFeatureManager featureManager;
+...
+Variant variant = await featureManager.GetVariantAsync(MyFeatureFlags.HelpText, CancellationToken.None);
+model.Text = variant.Configuration.Value;
+```
+
+For more details on Variants, see [here](https://learn.microsoft.com/en-us/azure/azure-app-configuration/feature-management-dotnet-reference#variants).
+
+#### Variant Service Provider
+
+Variant feature flags can be used in conjunction with dependency injection to surface different implementations of a service for different users. This is accomplished by using the Variant Service Provider.
+
+For more details on Variant Service Provider, see [here](https://learn.microsoft.com/en-us/azure/azure-app-configuration/feature-management-dotnet-reference#variants-in-dependency-injection)
+
+#### Telemetry
+
+Telemetry provides observability into flag evaluations, offering insights into which users received specific flag results. This enables more powerful metric analysis, such as experimentation.
+
+For more details on Telemetry, see [here](https://learn.microsoft.com/en-us/azure/azure-app-configuration/feature-management-dotnet-reference#telemetry).
+
+#### Microsoft Feature Management Schema
+
+Added support for variant feature flags defined using [Microsoft Feature Management schema](https://github.com/microsoft/FeatureManagement/blob/main/Schema/FeatureManagement.v2.0.0.schema.json). Variants and telemetry can be declared using [Microsoft Feature Flag schema v2](https://github.com/Azure/AppConfiguration/blob/main/docs/FeatureManagement/FeatureFlag.v2.0.0.schema.json). Here is a [Sample](https://github.com/microsoft/FeatureManagement-Dotnet/blob/main/examples/VariantAndTelemetryDemo/appsettings.json#L12).
+
+### Breaking Changes
+
+1. The constructor of `FeatureTagHelper` has been adjusted to take both an `IFeatureManager` and an `IVariantFeatureManager` for backwards compatability. Most use cases don't directly use the constructor, but if the constructor is directly called, the call will need to be adjusted to pass the additional parameter.
```

Comment:
- There is only one item. It's odd to have an index of 1.
- We are saying we added a breaking change "for backwards compatibility". It doesn't make sense.
- I still don't grasp the breaking change. Can we add before and after?

Reply:
```csharp
IFeatureManager featureManager = serviceProvider.GetRequiredService<IFeatureManager>();
FeatureTagHelper a = new FeatureTagHelper(featureManager);
```

Has become
```csharp
FeatureManager featureManager = serviceProvider.GetRequiredService<FeatureManager>();
FeatureTagHelper a = new FeatureTagHelper(featureManager, featureManager); 
// Or the more explicit
IFeatureManager featureManager = serviceProvider.GetRequiredService<IFeatureManager>();
IVariantFeatureManager variantFeatureManager = serviceProvider.GetRequiredService<IVariantFeatureManager>();
FeatureTagHelper a = new FeatureTagHelper(featureManager, variantFeatureManager);
```

This makes it backwards compatible with a custom `IFeatureManager`. While the variant logic it now has uses the `IVariantFeatureManager`.

Reply:
> There is only one item. It's odd to have an index of 1.

I prefer the index because it makes it clear that there is only a single breaking change. A paragraph sounds like there might be multiple. Could change the header to be singular but I prefer consistency. 

> We are saying we added a breaking change "for backwards compatibility". It doesn't make sense.

Unfortunately it's what's happening. We can leave it out if it makes it hard to read/understand.

Reply:
It may give you the opposite of what you intend. With an index of 1, it confuses your readers and they think you forgot to write 2 and more. You can use a bullet point instead.

The `FeatureTagHelper` belongs to the MS.FM.AspNetCore package, so this is the wrong place to add this note.

- What's the scenario one will call the constructor directly?
- The feature tag can now be controlled by variant feature flags. I will certainly list this under enhancement. We are not making this change for the backward compatibility. Didn't we have a document for this?

Reply:
> What's the scenario one will call the constructor directly?

Searching github shows a few results of people manually creating it for tests. I'm unaware of any other reason to call it directly. 

> The feature tag can now be controlled by variant feature flags. I will certainly list this under enhancement.

It was already released in preview4 and was a breaking change for customers using `IFeatureManager`- because it was adjusted to take `IVariantFeatureManager` instead. In this release- we're adjusting it to take both. This means customers using custom IFeatureManagers are still able to use their custom interface. However, it continues to be a breaking change from v3 since the constructor is changing.

I've been trying to call out the breaking change from `preview4` to `stable`. We added the extra parameter to support what we had in preview3 and before. (backwards compatibility that was lost on preview4). I can instead call out the breaking change from v3 -> v4 since this is kind of a recap of the preview changes.

How about:

MS.FM.AspNetCore

### Enhancements

#### FeatureTagHelper variants

FeatureTagHelper now supports variants. This allows MVC views to use the `<feature>` tag to conditionally render content based on whether a specific variant of a feature is assigned.

``` HTML+Razor
<feature name="FeatureX" variant="Alpha">
  <p>This can only be seen if variant 'Alpha' of 'FeatureX' is assigned.</p>
</feature>
```

For more details on ASP.NET views and variants, see [here](https://learn.microsoft.com/en-us/azure/azure-app-configuration/feature-management-dotnet-reference#view).

### Breaking Change

* The constructor for `FeatureTagHelper` now requires both an `IFeatureManager` and an `IVariantFeatureManager`. While this class is typically not instantiated directly, any direct instantiation will need to be updated.

Reply:
Most customers won't try the preview releases or track the breaking changes in preview releases. Therefore, for every stable release, we should include (repeat) all enhancements, bug fixes and breaking changes. We shouldn't force customers to read all the notes of preview releases. There is no way to tell whether something makes it to the final stable release or not if we don't call them out explicitly.

Please include everything since the last stable release if there is anything else missing.